src/hlsBinaries.ts

```diff
@@ -249,39 +249,64 @@ export async function findHaskellLanguageServer(
     // we manage HLS, make sure ghcup is installed/available
     await getGHCup(context, logger);
 
-    // get a preliminary hls wrapper for finding project GHC version,
-    // later we may install a different HLS that supports the given GHC
-    let wrapper = await getLatestWrapperFromGHCup(context, logger).then((e) =>
-      !e
-        ? callGHCup(context, logger, ['install', 'hls'], 'Installing latest HLS', true).then(() =>
-            callGHCup(
-              context,
-              logger,
-              ['whereis', 'hls'],
-              undefined,
-              false,
-              (err, stdout, _stderr, resolve, reject) => {
-                err ? reject("Couldn't find latest HLS") : resolve(stdout?.trim());
-              }
-            )
-          )
-        : e
-    );
+    // get a preliminary toolchain for finding the correct project GHC version (we need HLS and cabal/stack and ghc as fallback),
+    // later we may install a different toolchain that's more project-specific
+    const installGHC  = !executableExists('ghc');
+    let latestHLS     = await getLatestToolFromGHCup(context, logger, 'hls');
+    let latestCabal   = await getLatestToolFromGHCup(context, logger, 'cabal');
+    let latestStack   = await getLatestToolFromGHCup(context, logger, 'stack');
+    let recGHC        = await getLatestAvailableToolFromGHCup(context, logger, 'ghc', 'recommended');
```

Comment:
`getLatestToolFromGHCup` considers installed versions, `getLatestAvailableToolFromGHCup` only *installable* versions from ghcup.

We treat ghc a bit different here, because it's the most expensive download. So if any `ghc` is in PATH, we allow this version just for the bootstrapping procedure.